src/main/scala/com/fulcrumgenomics/bam/ConsensusCaller.scala

```diff
@@ -0,0 +1,369 @@
+                                  minConsensusBaseQuality: PhredValue     = DefaultMinConsensusBaseQuality,
+                                  minReads: Int                           = DefaultMinReads,
+                                  minMeanConsensusBaseQuality: PhredValue = DefaultMinMeanConsensusBaseQuality
+                                 )
+
+object ConsensusCaller {
+  /** Creates a consensus read from the given read and quality string tuples.  If no consensus read was created, None is returned. */
+  def fromBasesAndQualities(basesAndQualities: Seq[(String, String)],
+                            options: ConsensusCallerOptions = new ConsensusCallerOptions()
+                           ): Option[(String, String)] = {
+    if (basesAndQualities.exists { case (b, q) => b.length != q.length }) {
+      throw new IllegalArgumentException("Found a read with the bases and qualities having different length")
+    }
+    // check to see if we have enough reads.
+    if (basesAndQualities.length < options.minReads) return None
+
+    // extract the bases and qualities, and adjust the qualities based on the given options.
+    val bases: Seq[String] = basesAndQualities.map { _._1.toUpperCase }
+    val quals: Seq[Seq[PhredValue]] = basesAndQualities.map(bq => SAMUtils.fastqToPhred(bq._2))
+      .map { bq =>
+        adjustBaseQualities(
+          quals            = bq.map(q => PhredValue(q.toDouble)),
+          maxBaseQuality   = options.maxBaseQuality,
+          baseQualityShift = options.baseQualityShift,
+          errorRatePostUmi = options.errorRatePostUmi
+        )
+      } // NB: binary Phred, not ASCII
+
+    // get the most likely consensus bases and qualities
+    val (consensusBases, consensusQualities) = consensusCalls(
+      bases                   = bases,
+      quals                   = quals,
+      errorRatePreUmi         = options.errorRatePreUmi,
+      minConsensusBaseQuality = options.minConsensusBaseQuality,
+      minReads                = options.minReads
+      )
+
+    // check that the mean base quality is high enough
+    if (PhredValue.mean(consensusQualities:_*) < options.minMeanConsensusBaseQuality) None
+    else Some((consensusBases, SAMUtils.phredToFastq(consensusQualities.map(_.toByte).toArray[Byte])))
+  }
+
+  /** Creates a consensus read from the given records.  If no consensus read was created, None is returned. */
+  def fromSamRecords(records: Seq[SAMRecord],
+                     options: ConsensusCallerOptions = new ConsensusCallerOptions()
+                    ): Option[(String, String)] = {
+    // TODO: if mapped, this gets the bases in genomic order and not sequencing order.  Does it matter?
+    val bases = records.map { rec => rec.getReadString.toUpperCase }
+    val quals = records.map { rec => rec.getBaseQualityString }
+
+    fromBasesAndQualities( basesAndQualities=bases.zip(quals), options = options)
```

Comment:
This will be a little inefficient when processing BAM as I believe it will force base qualities to go:
`BAM-6bit -> Array[Byte] numeric Phred -> String -> Array[Byte] numeric Phred -> PhredValue`
for each record.